


# Bioinformatics Protocols at Kothary Lab

The following repository contains the protocols and logs of bioinformatics use in the Kothary Lab. Includes installation processes, pipeline configurations, test logs and outputs of bioinformatic processes/experiments run. Main contents is RNASeq analysis, conducted on Graham Cluster from DRAC.

Author: Majd Al-Aarg

Date of creation: August 21, 2023

Contact: majd.alaarg@gmail.com 


## Additional help

If you run into major issues or require help, you can consult the [nf-core documentation](https://nf-co.re/docs/usage/introduction) (click on link or just Google nf-core documentation)

If the pipeline fails, check the  [troubleshooting docs](https://nf-co.re/docs/usage/troubleshooting/)  and ask for help on the nf-core Slack channel for that particular pipeline (see  [https://nf-co.re/join](https://nf-co.re/join)).

### Acknowledgment
The rnaseq protcol in this repository are based on a [guide from the Ottawa Bioinformatics Core Facility](https://gitlab.com/ohri/obcf-documentation/-/blob/master/Files/nfcore-rnaseq-graham-HOWTO.md?ref_type=heads) that I redeveloped specifically for the Kothary Lab operations/experiments and to account for the changes in the Nextflow nf-core system at the time.


# Table of Contents (RNASeq)
1. [Downloading Data from The Centre for Applied Genomics Portal (TCAG)](https://github.com/majd-alaarg/Bioinformatic-Protocols-at-Kothary-Lab/blob/1a8fcb7e59a4c5da75ac4fa32cc6ee53c7118d85/RNASeq/Guides/Downloading%20Data%20from%20The%20Centre%20for%20Applied%20Genomics%20Portal%20(TCAG).pdf)
2. [Getting Started: Installing nf-core and RNASeq pipeline](https://github.com/majd-alaarg/Bioinformatic-Protocols-at-Kothary-Lab/blob/main/RNASeq/Guides/Getting%20Started%3A%20Installing%20nf-core%20and%20RNASeq%20pipeline.md)
3. [Running RNASeq pipeline](https://github.com/majd-alaarg/Bioinformatic-Protocols-at-Kothary-Lab/blob/main/RNASeq/Guides/Running%20RNASeq%20pipeline.md)
